### PR TITLE
Allow scalars to be used in setting empty columns

### LINKED ISSUE #17078: Inconsistent behavior in assigning constant value to a column in case of empty table
### Description

When assigning a constant value to a column, e.g. `table['col2'] = 1.0`, most of the time it works as expected, that the constant value is broadcasted to the column. But in the case that the table is empty and the column is a new one to be added, a `TypeError: Empty table cannot have column set to scalar value` is raised.

The behavior for statements such as `table['col2'] = 1.0`: 

|                 | Existing column | New Column |
| --------------- | --------------- | ---------- |
| Non empty table | ok              | ok         |
| empty table     | ok              | `TypeError`  |

**Workaround:** assigning with an explicit array, e.g., `table['col2'] = np.full(len(table), 1.0)`, would work for both empty and non-empty tables.

### Expected behavior

I'd expect `table['col2'] = 1.0` work even if the table is empty. 

This might not be considered a bug, as the behavior seems to be a deliberate choice in response to issue #3811 . I feel the inconsistency in the behavior is error prone, however. 

As a reference point, for `pandas` , the similar `data_frame['col2'] = 1.0` works in all cases.

### How to Reproduce

The snippets demonstrate the behavior described in the table above.

```python

from astropy.table import Table

#
# cases non-empty table: all works

# okay - replace a  column by broadcasting a constant
tab = Table(data={"col1": [1.0, 2.0], "col2": [9, 8]})
tab["col1"] = 1.3
print(tab)

# okay - creating a new column by broadcasting a constant
tab = Table(data={"col1": [1.0, 2.0], "col2": [9, 8]})
tab["col3"] = 3.3
print(tab)

#
# cases empty table: creating a new column fails

# okay - replace a column by broadcasting a constant
tab = Table(data={"col1": [], "col2": []})  # empty table
tab["col1"] = 1.3
print(tab)

# failed - create a new column by broadcasting a constant
tab = Table(data={"col1": [], "col2": []})  # empty table
tab["col3"] = 3.3  # TypeError
print(tab)

```

In the last case, creating a new column on an empty table, the `TypeError: Empty table cannot have column set to scalar value` is raised.

### Versions

```python
import astropy
try:
    astropy.system_info()
except AttributeError:
    import platform; print(platform.platform())
    import sys; print("Python", sys.version)
    import astropy; print("astropy", astropy.__version__)
    import numpy; print("Numpy", numpy.__version__)
    import erfa; print("pyerfa", erfa.__version__)
    try:
        import scipy
        print("Scipy", scipy.__version__)
    except ImportError:
        print("Scipy not installed")
    try:
        import matplotlib
        print("Matplotlib", matplotlib.__version__)
    except ImportError:
        print("Matplotlib not installed")
```
```
Windows-10-10.0.22631-SP0
Python 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:40:08) [MSC v.1938 64 bit (AMD64)]
astropy 6.1.3
Numpy 2.1.1
pyerfa 2.0.1.4
Scipy not installed
Matplotlib not installed
```

### PR DESCRIPTION
**Pull request by mhvk**
In #17078, it was noted that one should really be able to set a column with a scalar also when the size of the table is 0: just broadcast it to zero. This PR implements it. As a result, a number of tests had to be adjusted, and the one for the registry are arguably not great. So, opening as draft -- will add changelog entries later, once we decided what is best.

Fixes #17078

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/table/mixins/tests/test_registry.py [hunk 1]
```diff
@@ -74,10 +74,28 @@ def test_get_mixin_handler_str():
     assert get_mixin_handler(FULL_QUALNAME) is handle_spam
 
 
-def test_add_column():
+def test_add_column_to_empty_table():
     t = Table()
-    with pytest.raises(TypeError):
-        t["a"] = SpamData()
+    t["a"] = SpamData()
+    # Not what we want, really...
```
Comment (mhvk) [inline astropy/table/mixins/tests/test_registry.py:80]: This is the change I feel is a bit sad. But perhaps unavoidable?

File: astropy/table/mixins/tests/test_registry.py [hunk 2]
```diff
@@ -74,10 +74,28 @@ def test_get_mixin_handler_str():
...
+    register_mixin_handler(FULL_QUALNAME, handle_spam)
+    t = Table()
+    t["a"] = SpamData()
+
+    assert len(t) == 5
+    assert isinstance(t["a"], SpamWrapper)
+    assert_equal(t["a"].data, [0, 1, 3, 4, 5])
+
+
+def test_add_column_to_existing_table():
```
Comment (mhvk) [inline astropy/table/mixins/tests/test_registry.py:93]: This test would pass also on main, i.e., one would not reliably get an error for `t["a"] = SpamData()`.

**Review by mhvk — COMMENTED** [review]
Two inline comments to help review.

On using `np.broadcast_to` and `col._apply(np.broadcast_to)` - I think that is actually OK - I do not see how we can (or should) provide a general broadcasting mechanism for each possible mixin type, and if we do not know how to broadcast to larger length, I do not think we have to try to support broadcasting to zero length either. (Or, at least, let this be for follow-up!).

**Comment by mhvk** [comment]
@taldcroft - the doctest failure is for the same reason as that in `test_registry` - that the only reason that `t['data'] = ExampleDataClass()` failed before is that `t` is empty and we did not allow setting with a scalar object column.  I can fix this once we have decided what the best path forward is.

**Comment by taldcroft** [comment]
This has a problem which does not occur in my clunkier version :smile:.
```
In [2]: t = Table()

In [3]: t["a"] = 1.5

In [4]: t["a"]
Out[4]: <Column name='a' dtype='float64' length=0>

In [5]: t["a"] = 1.5

In [6]: t["a"]
Out[6]: 1.5

In [7]: t["a"].shape
Out[7]: ()

In [8]: t["b"] = 1
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File ~/git/astropy/astropy/table/table.py:2277, in Table.__len__(self)
   2276 try:
-> 2277     return len(OrderedDict.__getitem__(self.columns, self._first_colname))
   2278 except (AttributeError, KeyError):

AttributeError: 'Table' object has no attribute '_first_colname'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In[8], line 1
----> 1 t["b"] = 1

File ~/git/astropy/astropy/table/table.py:2136, in Table.__setitem__(self, item, value)
   2132 def __setitem__(self, item, value):
   2133     # If the item is a string then it must be the name of a column.
   2134     # If that column doesn't already exist then create it now.
   2135     if isinstance(item, str) and item not in self.colnames:
-> 2136         self.add_column(value, name=item, copy=True)
   2138     else:
   2139         n_cols = len(self.columns)

File ~/git/astropy/astropy/table/table.py:2442, in Table.add_column(self, col, index, name, rename_duplicate, copy, default_name)
   2434 # For scalars and for arrays with length 1, allow broadcasting to the
   2435 # length of the table. This includes zero-length tables, i.e., we
   2436 # broadcast the column to zero length (since astropy 7.0; this
   (...)
   2439 # which for a scalar will be length zero (which is most easily achieved
   2440 # by pass-through here).
   2441 if col.shape == () or (col.shape[0] == 1 and self.columns):
-> 2442     new_shape = (len(self),) + getattr(col, "shape", ())[1:]
   2443     if isinstance(col, np.ndarray):
   2444         col = np.broadcast_to(col, shape=new_shape, subok=True)

File ~/git/astropy/astropy/table/table.py:2284, in Table.__len__(self)
   2282 # Get the first column name
   2283 self._first_colname = next(iter(self.columns))
-> 2284 return len(self.columns[self._first_colname])

TypeError: len() of unsized object
```

**Comment by mhvk** [comment]
Actually, that is a more general bug, unrelated to my changes:
```
qt = QTable()
qt["a"] = np.arange(3.)
qt["a"] = 0.
qt
TypeError: len() of unsized object
```
(this is on main down to 5.2.1, so old). Just goes to show that one should really try to avoid to do the same check in multiple places (here, `add_column` and `replace_column`)...

**Comment by taldcroft** [comment]
> Actually, that is a more general bug, unrelated to my changes:

Ugh...

**Comment by mhvk** [comment]
Fix for that is in #17105. I now based this PR on it, and added the example you gave in https://github.com/astropy/astropy/pull/17102#issuecomment-2386767337 as a test case.

The main question, though, is whether it is OK to have non-registered classes be included as object arrays by default, i.e., registration allows them to be treated correctly, but one cannot get an error if they just look like a random object that gets turned into a column with `dtype=object` (I don't really see a way to avoid it, sadly, and, as noted, it already happened if there was another column in the table, so it is not *that* big of a change).

**Comment by taldcroft** [comment]
@mhvk - I spent some time looking at this and it seems fine to move forward. I think the `SpamData()` stuff is doing what it should for an unregistered object so that seems OK. The tests look about right.

About the `broadcast_to`, is it too early to try checking if the object implements the array API with `__array_namespace__` and then using the `broadcast_to` function from that? Could `ShapedLikeNDArray` provide `__array_namespace__` as access to all the numpy functions currently implemented via `_apply`?

**Comment by mhvk** [comment]
> About the `broadcast_to`, is it too early to try checking if the object implements the array API with `__array_namespace__` and then using the `broadcast_to` function from that? Could `ShapedLikeNDArray` provide `__array_namespace__` as access to all the numpy functions currently implemented via `_apply`?

That's a really neat idea! But better as a new issue, I think, since I do not quite feel confident about what it would actually involve... See #17136.

**Review by taldcroft — APPROVED** [review]
Looks good, thanks! By crazy coincidence, I ended up wanting this fix twice last week...

**Pull request merged**